From 8ffb2c233432024c7adb24a1f7a0b25b28112014 Mon Sep 17 00:00:00 2001
From: Benjamin Gilbert <bgilbert@cs.cmu.edu>
Date: Tue, 7 Nov 2023 10:55:08 -0600
Subject: [PATCH] meson: add a private option to filter out non-OpenSlide
 symbols

On macOS, openslide-bin needs to use a linker option to filter out
non-OpenSlide exported symbols.  It can't set the option globally because
that would break OpenSlide Java, and can't set a per-project option
because Meson doesn't support that.  Add a private option to do the
filtering until Meson provides syntax for this.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>
---
 meson.build       | 10 ++++++++++
 meson_options.txt |  6 ++++++
 src/meson.build   |  1 +
 3 files changed, 17 insertions(+)

diff --git a/meson.build b/meson.build
index 0b4d7d47b695..852b976419a9 100644
--- a/meson.build
+++ b/meson.build
@@ -161,6 +161,16 @@ endif
 
 # Test suite options
 visibility = get_option('_export_internal_symbols') ? '' : 'hidden'
+libopenslide_link_args = (
+  # On macOS, openslide-bin needs to keep non-OpenSlide symbols out of the
+  # combined dylib, but it can't do it via global link options without
+  # breaking the exports of the OpenSlide Java JNI library.  Until Meson
+  # allows passing compiler flags to individual subprojects, provide a
+  # workaround here.
+  # https://github.com/mesonbuild/meson/issues/11002
+  get_option('_filter_external_symbols') and host_machine.system() == 'darwin'
+  ? ['-Wl,-exported_symbol,_openslide_*'] : []
+)
 if get_option('_gcov')
   add_project_arguments(
     '-O0',
diff --git a/meson_options.txt b/meson_options.txt
index 1a71bf92c235..a11494c85f7a 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -29,6 +29,12 @@ option(
   value : false,
   description : 'For test suite; do not use',
 )
+option(
+  '_filter_external_symbols',
+  type : 'boolean',
+  value : false,
+  description : 'For openslide-bin; do not use',
+)
 option(
   '_gcov',
   type : 'boolean',
diff --git a/src/meson.build b/src/meson.build
index f9d184be1bc7..e8997cae00ed 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -78,6 +78,7 @@ libopenslide = library('openslide',
   openslide_sources,
   version : soversion,
   c_args : ['-D_OPENSLIDE_BUILDING_DLL', '-DG_LOG_DOMAIN="OpenSlide"'],
+  link_args : libopenslide_link_args,
   gnu_symbol_visibility : visibility,
   include_directories : config_h_include,
   dependencies : [
-- 
2.41.0

